proposals/Hole Punching.md

```diff
@@ -0,0 +1,146 @@
+
+Also, based on anecdotal evidence in the wild and engineering war stories (see [Tailscales’s blog](https://tailscale.com/blog/how-nat-traversal-works/)), **Cone NATs are much more pervasive in Home ISPs over Symmetric NATs, which justifies the~60-80% success for Hole Punching**.
+
+Given that libp2p is a library to build peer to peer applications,**implementing Hole Punching in libp2p will allow any application/network that builds on top of libp2p to also benefit from Hole Punching**. This will be a significant step forward in the ease of building well connected p2p networks and will be a major added motivation for developers to build on top of our stack. Based on our research, **no such library that comes with out of the box NAT traversal via Hole Punching exists out in the wild today and we have the opportunity to provide the first such library & infra to herald the age of better connectivity in Web3 apps**.
+
+### Assumptions & hypotheses
+
+*   Applications that build on top of our stack want peers to be _directly_ reachable from the network even though they are behind a NAT (~80% peers in the current DHT network).
+*   PL is not willing to keep funding expensive  bandwidth-unrestricted Relay servers as the network keeps growing to enable data transfer to/from NATT’d peers.
+*   Users would love to use our p2p stack if doing so means the applications they build get NAT traversal via Hole Punching out of the box.
+*   Enabling better connectivity in IPFS, Filecoin and other applications that build on top of Libp2p in a world suffering from the tyranny of NATs is a success metric/priority for us.
+
+### User workflow example
+
+The User builds an application such as IPFS, Filecoin, Ethereum etc. on top of Libp2p and gets _direct_ connectivity between NATT’d peers (albeit ~60% for TCP & ~80% for QUIC) out of the box without bringing up additional infrastructure and with minimal configuration.
+
+### Impact
+🔥🔥🔥
+
+*   Libp2p will be one of the first libraries in the web3 ecosystem that provides Hole Punching and hence better connectivity out of the box. This will increase the functionality of our stack, and will encourage more developers to build on top of it.
+*   PL’s applications such as Filecoin & IPFS will get turbocharged with better connectivity as they build on top of libp2p.
+*   Important projects such as Eth2, 0x etc that build on top of libp2p will ALSO get this huge benefit should they choose to use it. Eth2 is likely to need this feature in Phase 2, which introduces browser-based light clients.
+*   New browser-centric use cases will be possible when this functionality is implemented in js-libp2p.
```

Comment:
Could this be elaborated a bit more, I am not sure I understand what are those use cases or how hole punching would make browser nodes dialable.

Reply:
@Gozala 

Given that JS peers can't connect to go peers directly right now, shipping the Limited Relay based decentralised signalling is the first step in implementing a WebRTC transport in go & js libp2p where signalling does NOT rely on centralised STAR servers.

Reply:
"JS peers can't connect to go peers directly right now"

The root reason for this is that `go-ipfs` does not have the `websockets` transport by default. Also these nodes are not usually reachable with a SSL+DNS multiaddr, which is essential for the browser.

Unless we plan to have `webRTC` used by default in `go-ipfs`, the browser limitations will not be solved.

I remember that Chrome had speed limitations for `webRTC`, which made websockets a faster option. @Gozala do you know if this is still the case? If this is the case, I think it would be better to focus on a solution for generating certificates instead of `webRTC` in go. (and of course enable websockets by default in go)

Reply:
> I remember that Chrome had speed limitations for `webRTC`, which made websockets a faster option. @Gozala do you know if this is still the case? If this is the case, I think it would be better to focus on a solution for generating certificates instead of `webRTC` in go. (and of course enable websockets by default in go)

I do not know what speed limitations chrome or other browsers put on WebRTC, but there are problems with WebRTC beyond that:

1. WebRTC is not implemented in workers threads.
   - This makes building responsive apps a lot more challenging given all the other worker that is happening on main thread
   - Prevents IPFS node sharing a node across contexts like tabs or iframes.
2. I have anecdotal reports that in practice WebRTC is impractical without TURN servers, suggesting that often times data is relayed anyway.
3. Again anecdotal, evidence suggests that WebRTC seems to cause significant CPU load

All the above combined often leads teams to pivot towards WebSocket based solutions. I also heard teams reporting reduced bills when operating WebSocket based relay as opposed to TURN servers.

Please take all this with a grain of last, because I have not seen any comprehensive studies to support anecdotal evidence.

Reply:
And here is some data regarding ICE success rate https://medium.com/the-making-of-whereby/so-i-read-that-20-of-webrtc-calls-fail-67b185e49765